File: beakjun/bfs_dfs/multi_case.js
const readFileSyncAddress =
  process.platform === "linux" ? "/dev/stdin" : "example.txt";
let data = require("fs")
  .readFileSync(readFileSyncAddress)
  .toString()
  .split("\n");

let input = [];
let M = 0;
let N = 0;

const main = () => {
  const graph = [];
  const visited = [];

  for (let i = 0; i < N; i++) {
    graph[i] = input[i].split(" ").map(Number);
    visited[i] = new Array(M).fill(false);
  }

  const bfs = (yPos, xPos) => {
    const xMove = [0, 0, -1, 1, -1, -1, 1, 1];
    const yMove = [1, -1, 0, 0, 1, -1, 1, -1];
    const queue = [];
    queue.push({ yPos: yPos, xPos: xPos });
    visited[yPos][xPos] = true;

    while (queue.length) {
      const { yPos, xPos } = queue.shift();
      for (let i = 0; i < 8; i++) {
        const nextY = yPos + yMove[i];
        const nextX = xPos + xMove[i];
        if (nextY >= 0 && nextY < N && nextX >= 0 && nextX < M) {
          if (!visited[nextY][nextX] && graph[nextY][nextX]) {
            visited[nextY][nextX] = true;
            queue.push({ yPos: nextY, xPos: nextX });
          }
        }
      }
    }
  };

  let count = 0;
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (!visited[i][j] && graph[i][j]) {
        count++;
        bfs(i, j);
      }
    }
  }
  console.log(count);
};

data.forEach((line) => {
  if (!N) {
    if (line === "0 0") {
      return;
    }
    [M, N] = line.split(" ").map(Number);
  } else {
    input.push(line);
    if (input.length === N) {
      main();
      N = 0;
      input = [];
    }
  }
});

File: beakjun/bfs_dfs/multi_case.py
import sys
from collections import deque
input = sys.stdin.readline

n,m = map(int,input().split())
graph = []
for i in range(n):
    graph.append(list(map(int,input().split())))


dx = [1,-1,0,0]
dy = [0,0,1,-1]
ans = []

pic_count =0
max_count = 0
visited = [[0]*m for _ in range(n)]

def bfs(graph,x,y):
    q= deque()
    q.append([x,y])
    graph[x][y] =0
    cnt =1
    
    while q:
        x,y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0<=nx<n and 0<=ny<m and visited[nx][ny]==0:
                if graph[nx][ny] == 1:
                    graph[nx][ny] = 0
                    q.append([nx,ny])
                    cnt += 1
    return cnt

paint = []
for i in range(n):
    for j in range(m):
        if(graph[i][j] == 1):
            paint.append(bfs(graph,i,j))
            
if len(paint) == 0:
    print(len(paint))
    print(0)
else:
    print(len(paint))
    print(max(paint))